lib/sensors/GPS.cpp

```diff
@@ -0,0 +1,494 @@
+}
+void GPS::addToChecksum(uint8_t incoming) {
+	m_rollingChecksumA += incoming;
+	m_rollingChecksumB += m_rollingChecksumA;
+}
+
+void GPS::setSerialRate(uint32_t baudrate, uint8_t uartPort) {
+	getPortSettings(uartPort); //load payloadCfg with current port settings
+	m_packetCfg.cls = UBX_CLASS_CFG;
+	m_packetCfg.id = UBX_CFG_PRT;
+	m_packetCfg.len = 20;
+	m_packetCfg.startingSpot = 0;
+
+	//change necessary msg data
+	payloadCfg[8] = baudrate;
+	payloadCfg[9] = baudrate >> 8;
+	payloadCfg[10] = baudrate >> 16;
+	payloadCfg[11] = baudrate >> 24;
+
+	sendSerialCommand(&m_packetCfg);
+}
+
+//Configure a given port to output UBX, NMEA, and/or RCTM3
+//Port:     	use COM_PORT_ constants
+//Settings: 	use COM_TYPE_ constants
+void GPS::setPortOutput(uint8_t portID, uint8_t outStreamSettings) {
+	getPortSettings(portID); //load payloadCfg with current port settings
+	m_packetCfg.cls = UBX_CLASS_CFG;
+	m_packetCfg.id = UBX_CFG_PRT;
+	m_packetCfg.len = 20;
+	m_packetCfg.startingSpot = 0;
+
+	//change necessary msg data
+	payloadCfg[14] = outStreamSettings;
+	sendSerialCommand(&m_packetCfg);
+}
+
+//Configure a given port to accept UBX, NMEA, RTCM3 or a combination thereof
+//Port:     	use COM_PORT_ constants
+//Settings: 	use COM_TYPE_ constants
+void GPS::setPortInput(uint8_t portID, uint8_t inStreamSettings) {
+	getPortSettings(portID); //load payloadCfg with current port settings
+	m_packetCfg.cls = UBX_CLASS_CFG;
+	m_packetCfg.id = UBX_CFG_PRT;
+	m_packetCfg.len = 20;
+	m_packetCfg.startingSpot = 0;
+
+	//change necessary msg data
+	payloadCfg[12] = inStreamSettings;
+	sendSerialCommand(&m_packetCfg);
+}
+
+void GPS::setNavigationFrequency(uint8_t navFreq) {
+	m_packetCfg.cls = UBX_CLASS_CFG;
+	m_packetCfg.id = UBX_CFG_RATE;
+	m_packetCfg.len = 0;
+	m_packetCfg.startingSpot = 0;
+
+	//load payloadCFG with current settings
+	sendSerialCommand(&m_packetCfg);
+	if (!waitForAck(UBX_CLASS_CFG, UBX_CFG_RATE)) {
+		//TODO: handle failure case
+	}
+
+	uint16_t measurementRate = 1000 / navFreq;
+
+	//payloadCfg is now loaded with current bytes. Change only the ones we need to
+	payloadCfg[0] = measurementRate & 0xFF; //measRate LSB
+	payloadCfg[1] = measurementRate >> 8;   //measRate MSB
+
+	sendSerialCommand(&m_packetCfg);
+}
+
+void GPS::configureMessage(uint8_t msgClass, uint8_t msgID, uint8_t portID, uint8_t sendRate) {
+	m_packetCfg.cls = UBX_CLASS_CFG;
+	m_packetCfg.id = UBX_CFG_MSG;
+	m_packetCfg.len = 8;
+	m_packetCfg.startingSpot = 0;
+
+	//Clear packet payload
+	for (int x = 0; x < m_packetCfg.len; x++) {
+		m_packetCfg.payload[x] = 0;
+	}
+	m_packetCfg.payload[0] = msgClass;
+	m_packetCfg.payload[1] = msgID;
+	//Send rate is how many cycles pass between each send
+	// For example, if the rate of a navigation message is set to 2, the message is sent every 2nd navigation solution.
+	m_packetCfg.payload[2 + portID] = sendRate;
+
+	sendSerialCommand(&m_packetCfg);
+}
+
+void GPS::saveConfiguration()
+//saves config in receiver battery backed RAM
```

Comment:
Is this storage internal to the chip itself?

Reply:
Yes

Reply:
Is it NVRAM? `saveConfigToInternalNVRAM`?

Reply:
I'm not entirely sure if its flash memory or battery backed RAM, will need to test with the actual unit first. I will update the name.